Change: make lists in savegames consistent

TrueBrain opened the pull request:
## Motivation / Problem

We have a bunch of different lists, which are all done slightly different.

- We have a "fixed-length" list, where the code dictates the length. No validation is done if that same value was also used when writing to disk, which means we have plenty of savegame bumps for cases where it did.
- We have lists like `SL_REFLIST` that have a 32bit length indicator.
- We have all kind of custom stuff with `SL_STRUCTLIST`, where sometimes the length is stored in the parent blob, saved/loaded to a global. But sometimes totally different solutions are used.

This makes self-described savegames really hard (the ultimate goal of this set of PRs, see #9322), but also means savegame-readers need to be really aware of what they are reading, and support a lot of different savegame-versions to do the right thing.

## Description

So .. as you might expect, I set out to fix that problem. Now the rules are simple:

- Any list, no matter who/what you are, has a gamma before the list to indicate how many entries are in the list.

In most cases this means you know the length. For `SL_STRUCT` this is not the case, and as such you need to know what struct is stored to figure out the length. This will be addressed in a later PR.

As using `SL_STRUCT` is normally only useful if it is an optional field (why else bother making it in a new struct, and not just embed it in the parent?), internally it is a `SL_STRUCTLIST` of either 0 or 1 in length. So if it is empty (nothing saved), the length-field is 0. Otherwise, length-field is 1.

## Limitations

- A lot of testing is involved, see #9339 and #9322 for details. So no known issues or limitations.
- Other solutions have been considered, instead of a length-field, like for example a "type", where the last entry is an empty one indicating "end". But given the current state of the code, that is vastly more work.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TrueBrain committed:
Change: store length of SL_STRUCTLIST in the savegame

This wasn't consistently done, and often variables were used that
were read by an earlier blob. By moving it next to the struct
itself, the code becomes a bit more self-contained and easier to
read.

Additionally, this allows for external tooling to know how many
structs to expect, instead of having to know where to find the
length-field or a hard-coded value that can change at any moment.

TrueBrain committed:
Change: indicate in the savegame if a SL_STRUCT contains any data

This helps external tooling to understand if a SL_STRUCT should
be skipped when reading. Basically, this transforms an SL_STRUCT
into a SL_STRUCTLIST with either 0 or 1 length.

TrueBrain committed:
Change: switch SL_DEQUE and SL_REFLIST length field to a gamma

The current SaveLoad is a bit inconsistent how long a length field
is. Sometimes it is a 32bit, sometimes a gamma. Make it consistent
across the board by making them all gammas.

TrueBrain committed:
Change: prefix SL_ARR with the length of the array

This means that during loading we can validate that what is saved
is also that what is expected. Additionally, this makes all list
types similar to how they are stored on disk:
First a gamma to indicate length, followed by the data.
The size still depends on the type.

rubidium42 commented on src/saveload/company_sl.cpp:
```diff
@@ -395,12 +403,20 @@ class SlCompanyLiveries : public DefaultSaveLoadHandler<SlCompanyLiveries, Compa
+
+	void Load(CompanyProperties *c) const override
+	{
+		size_t num_liveries = IsSavegameVersionBefore(SLV_63) ? LS_END - 4 : IsSavegameVersionBefore(SLV_85) ? LS_END - 2: IsSavegameVersionBefore(SLV_SAVELOAD_LIST_LENGTH) ? (int)LS_END : SlGetStructListLength(LS_END);
```
Is the cast to int really necessary?

Maybe create a helper function that returns the liveries? That makes it quite a bit more legible overall.
```
static size_t GetNumberOfLiveriesInTheSavegame() {
    if (IsSavegameVersionBefore(SLV_63)) return LS_END - 4;
    if (IsSavegameVersionBefore(SLV_85)) return LS_END - 2;
    if (IsSavegameVersionBefore(SLV_SAVELOAD_LIST_LENGTH)) return LS_END;
    return SlGetStructListLength(LS_END);
}
```

rubidium42 commented on src/saveload/station_sl.cpp:
```diff
@@ -369,7 +383,7 @@ class SlStationGoods : public DefaultSaveLoadHandler<SlStationGoods, BaseStation
 		}
 
-		uint num_cargo = IsSavegameVersionBefore(SLV_55) ? 12 : IsSavegameVersionBefore(SLV_EXTEND_CARGOTYPES) ? 32 : NUM_CARGO;
+		size_t num_cargo = IsSavegameVersionBefore(SLV_55) ? 12 : IsSavegameVersionBefore(SLV_EXTEND_CARGOTYPES) ? 32 : IsSavegameVersionBefore(SLV_SAVELOAD_LIST_LENGTH) ? (size_t)NUM_CARGO : SlGetStructListLength(NUM_CARGO);
```
Maybe a helper function here as well to improve legibility?

rubidium42 commented on src/saveload/town_sl.cpp:
```diff
@@ -124,14 +124,15 @@ class SlTownSupplied : public DefaultSaveLoadHandler<SlTownSupplied, Town> {
 	void Load(Town *t) const override
 	{
-		uint num_cargo = IsSavegameVersionBefore(SLV_EXTEND_CARGOTYPES) ? 32 : NUM_CARGO;
+		size_t num_cargo = IsSavegameVersionBefore(SLV_SAVELOAD_LIST_LENGTH) ? (IsSavegameVersionBefore(SLV_EXTEND_CARGOTYPES) ? 32 : NUM_CARGO) : SlGetStructListLength(NUM_CARGO);
```
The previously mentioned helper for stations can probably be reused here as well, right?

rubidium42 commented on src/saveload/saveload.cpp:
```diff
@@ -1043,45 +1043,28 @@ static void SlStdString(void *ptr, VarType conv)
- * @param length The length of the array counted in elements
- * @param conv VarType type of the variable that is used in calculating the size
+ * Internal function to save/Load a list of SL_VARs.
+ * SlCopy() and SlArray() are very similar, with the exception of the header.
```
Just for confirmation:
SlCopy is "meant" for anything that has a dynamic size?
SlArray is "meant" for anything that has a fixed size?
Or am I missing something else here? I can't really figure out from the comments of either SlCopy or SlArray.
And I'm seeing SlCopies with fixed sizes, so I guess what I'm thinking is wrong. In any case, I guess it needs to be clearer whether when to use SlCopy or SlArray.

TrueBrain replied:
Will update the comments, but basically:

`SlCopy` you can use from outside `saveload.c`, `SlArray` is only usable via an `SL_ARR`. The only reason for the difference is that one has a length-field and the other doesn't.

TrueBrain committed:
Codechange: use SL_NULL if you want to store null-bytes or load to nothing

Using SL_ARR for this gives us a bit of trouble later on, where we
add a length-field to SL_ARR. This of course is not the intention
of SLE_CONDNULL. So better seperate it.

TrueBrain committed:
Codechange: make it more obvious SlArray supports SLE_VAR_NULL

In the end, the code was already doing the right thing, but a few
functions deep, and not really obvious. When validating what objects
can handle SLE_VAR_NULL, it is nicer to just have this obvious.

rubidium42 commented on src/saveload/station_sl.cpp:
```diff
@@ -344,15 +360,25 @@ class SlStationGoods : public DefaultSaveLoadHandler<SlStationGoods, BaseStation
+	 * Get the number of cargoes used by this savegame version.
+	 * @return The number of cargoes used by this savegame version.
+	 */
+	size_t GetNumCargo() const {
```
`{` on newline.

rubidium42 commented on src/saveload/town_sl.cpp:
```diff
@@ -122,16 +122,28 @@ class SlTownSupplied : public DefaultSaveLoadHandler<SlTownSupplied, Town> {
+	 * Get the number of cargoes used by this savegame version.
+	 * @return The number of cargoes used by this savegame version.
+	 */
+	size_t GetNumCargo() const {
```
`{` on newline.

rubidium42 commented on src/saveload/company_sl.cpp:
```diff
@@ -395,12 +403,32 @@ class SlCompanyLiveries : public DefaultSaveLoadHandler<SlCompanyLiveries, Compa
+	 * Get the number of liveries used by this savegame version.
+	 * @return The number of liveries used by this savegame version.
+	 */
+	size_t GetNumLiveries() const {
```
`{` on newline.
